processor/src/fast/mod.rs

```diff
@@ -0,0 +1,324 @@
+    }
+
+    pub fn execute(&mut self, program: &Program) -> Result<StackOutputs, ExecutionError> {
+        self.execute_mast_node(program.entrypoint(), program.mast_forest())?;
+
+        StackOutputs::new(self.stack.iter().rev().copied().collect())
+            .map_err(|_| ExecutionError::OutputStackOverflow(self.stack.len() - MIN_STACK_DEPTH))
+    }
+
+    fn execute_mast_node(
+        &mut self,
+        node_id: MastNodeId,
+        program: &MastForest,
+    ) -> Result<(), ExecutionError> {
+        let node = program
+            .get_node_by_id(node_id)
+            .ok_or(ExecutionError::MastNodeNotFoundInForest { node_id })?;
+
+        match node {
+            MastNode::Block(basic_block_node) => {
+                self.execute_basic_block_node(basic_block_node, program)
+            },
+            MastNode::Join(join_node) => {
+                self.execute_mast_node(join_node.first(), program)?;
+                self.execute_mast_node(join_node.second(), program)
+            },
+            MastNode::Split(split_node) => {
+                let condition = self.stack[0];
+                if condition == ONE {
+                    self.execute_mast_node(split_node.on_true(), program)
+                } else if condition == ZERO {
+                    self.execute_mast_node(split_node.on_false(), program)
+                } else {
+                    return Err(ExecutionError::NotBinaryValue(condition));
+                }
+            },
+            MastNode::Loop(_loop_node) => todo!(),
+            MastNode::Call(_call_node) => todo!(),
+            MastNode::Dyn(_dyn_node) => todo!(),
+            MastNode::External(_external_node) => todo!(),
+        }
+    }
+
+    fn execute_basic_block_node(
+        &mut self,
+        basic_block_node: &BasicBlockNode,
+        _program: &MastForest,
+    ) -> Result<(), ExecutionError> {
+        let mut triplet_op_iter = TripletIterator::new(basic_block_node.operations());
+
+        while let Some((op1, op2, op3)) = triplet_op_iter.next() {
+            match (op1, op2, op3) {
+                (Operation::Swap, Operation::Dup1, Operation::Add) => self.swap_dup1_add(),
+                _ => {
+                    // No macro operation found; execute the first operation, and push the other two
+                    // back at the front of the iterator
+                    self.execute_op(op1, _program)?;
+                    triplet_op_iter.push_front(op2, op3);
+                },
+            }
+        }
+
+        // Execute any remaining operations
+        {
+            let (first, second) = triplet_op_iter.remainder();
+            if let Some(op) = first {
+                self.execute_op(op, _program)?;
+            }
+            if let Some(op) = second {
+                self.execute_op(op, _program)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn execute_op(
+        &mut self,
+        operation: &Operation,
```

Comment:
I wonder how much of an impact it has that the `Operation` is pretty heavy. Specifically, I think memory footprint of a single instance is something like 16 bytes. Maybe the compiler can optimize this away somehow, but if not, loading an operation may require multiple memory reads.